pkg/gofr/datasource/mongo/mongo.go

```diff
@@ -108,18 +115,71 @@ func (c *Client) Connect() {
 	}
 
 	if err = m.Ping(ctx, nil); err != nil {
-		c.logger.Errorf("could not connect to mongoDB at %v due to err: %v", c.config.URI, err)
+		c.logger.Errorf("could not connect to MongoDB at %v due to err: %v", host, err)
 		return
 	}
 
-	c.logger.Logf("connected to mongoDB successfully at %v to database %v", c.config.URI, c.config.Database)
+	c.logger.Logf("connected to MongoDB successfully at %v to database %v", host, c.config.Database)
 
 	mongoBuckets := []float64{.05, .075, .1, .125, .15, .2, .3, .5, .75, 1, 2, 3, 4, 5, 7.5, 10}
-	c.metrics.NewHistogram("app_mongo_stats", "Response time of MONGO queries in milliseconds.", mongoBuckets...)
+	c.metrics.NewHistogram("app_mongo_stats", "Response time of MongoDB queries in milliseconds.", mongoBuckets...)
 
 	c.Database = m.Database(c.config.Database)
 
-	c.logger.Logf("connected to MongoDB at %v to database %v", uri, c.Database)
+	c.logger.Logf("connected to MongoDB at %v to database %v", host, c.Database)
+}
+
+func generateMongoURI(config *Config) (uri, host string, err error) {
+	if config.URI != "" {
+		host, err = getDBHost(config.URI)
+		if err != nil {
+			return "", "", err
+		}
+
+		return config.URI, host, nil
+	}
+
+	switch {
+	case config.Host == "":
+		return "", "", fmt.Errorf("%w: host is empty", errMissingField)
+	case config.Port == 0:
+		return "", "", fmt.Errorf("%w: port is empty", errMissingField)
+	case config.Database == "":
+		return "", "", fmt.Errorf("%w: database is empty", errMissingField)
+	}
+
+	u := &url.URL{
+		Scheme: "mongodb",
+		Host:   net.JoinHostPort(config.Host, strconv.Itoa(config.Port)),
+		Path:   "/" + url.PathEscape(config.Database),
+	}
+
+	if config.User != "" && config.Password != "" {
+		u.User = url.UserPassword(url.QueryEscape(config.User), url.QueryEscape(config.Password))
+	}
+
+	q := u.Query()
+	q.Set("authSource", "admin")
+	u.RawQuery = q.Encode()
+
+	return u.String(), u.Hostname(), nil
+}
+
+func getDBHost(uri string) (host string, err error) {
+	parsedURL, err := url.ParseRequestURI(uri)
+	if err != nil {
+		return "", err
```

Comment:
```suggestion
		return "", fmt.Errorf("%w: %w", errParseHost, err)
```
See https://github.com/gofr-dev/gofr/pull/1355#discussion_r1906559791

Reply:
I don't feel the same, 
If the url is empty....empty url error is already wrapped around `"error generating Mongo URI"`, errParseHost error seems irrelevant in that context.

Reply:
Indeed